clodius/tiles/fasta.py

```diff
@@ -0,0 +1,260 @@
+from pyfaidx import Fasta
+import re
+import numpy as np
+import functools as ft
+import pandas as pd
+import logging
+import base64
+
+logger = logging.getLogger(__name__)
+
+TILE_SIZE = 1024
+
+
+def get_quadtree_depth(chromsizes):
+    tile_size_bp = TILE_SIZE
+    min_tile_cover = np.ceil(sum(chromsizes) / tile_size_bp)
+    return int(np.ceil(np.log2(min_tile_cover)))
+
+
+
+def natsort_key(s, _NS_REGEX=re.compile(r"(\d+)", re.U)):
+    return tuple([int(x) if x.isdigit() else x for x in _NS_REGEX.split(s) if x])
+
+
+def natcmp(x, y):
+    if x.find("_") >= 0:
+        x_parts = x.split("_")
+        if y.find("_") >= 0:
+            # chr_1 vs chr_2
+            y_parts = y.split("_")
+
+            return natcmp(x_parts[1], y_parts[1])
+        else:
+            # chr_1 vs chr1
+            # chr1 comes first
+            return 1
+    if y.find("_") >= 0:
+        # chr1 vs chr_1
+        # y comes second
+        return -1
+
+    _NS_REGEX = re.compile(r"(\d+)", re.U)
+    x_parts = tuple([int(a) if a.isdigit() else a for a in _NS_REGEX.split(x) if a])
+    y_parts = tuple([int(a) if a.isdigit() else a for a in _NS_REGEX.split(y) if a])
+
+    # order of these parameters is purposefully reverse how they should be
+    # ordered
+    for key in ["m", "y", "x"]:
+        if key in y.lower():
+            return -1
+        if key in x.lower():
+            return 1
+
+    try:
+        if x_parts < y_parts:
+            return -1
+        elif y_parts > x_parts:
+            return 1
+        else:
+            return 0
+    except TypeError:
+        return 1
+
+
+def natsorted(iterable):
+    return sorted(iterable, key=ft.cmp_to_key(natcmp))
+
+
+def get_chromsizes(fapath):
+    with Fasta(fapath, one_based_attributes=False) as fa:
+        chromsizes = dict((seq, len(fa.records[seq])) for seq in fa.keys())
```

Comment:
Here you're reading the raw sequence for each chromosome. I'm not too familiar with `pyfaidx` but if you have a `.fai` index of a fasta file, you can read the sizes of each chromosome without reading their actual sequence (see the `samtools faidx` command).

Reply:
I'm aware that it's not necessary to read the sequence to get the chromsizes, but actually `fa.records[seq]` does not read the sequence; it only reads the index. To read the sequence you must access `fa.records[seq].seq`. This is possible due to clever Python tricks within `pyfaidx`, from my investigation into its source code.